LX-53 REST API improvements

pomegranited opened the pull request:
Provides several improvements to the REST API, detailed below.

**JIRA tickets**:  LX-53, SE-281

**Testing instructions**:

- [x] Add expand field to /bundles/ and /bundle_versions/ to inline the files list:  b25d9a6 
   * Visit http://localhost:8888/api/v1/bundle/ and view a bundle version.
   * Visit http://localhost:8888/api/v1/bundle_versions/ and view a bundle version.
   * Append `?expand=files` to the urls above, and note that the `files` hyperlink is converted to an array of files data, which matches what was shown by following the `files` hyperlink.
   * Appending `?expand=files&format=json` also works.
   * Appending `?expand=files` to the `/api/v1/bundle/` and `/api/v1/bundle_versions/` list views has no effect (previous to f47b553, it threw an error).
- [x] Add filters to the endpoints: c47b482
   * list Bundles associated with a Collection, e.g. 
      http://localhost:8888/api/v1/bundles/?collection__uuid=4ffa83df9b93499e91c29f5a7ad9
   * list BundleVersions for a Bundle, e.g.
      http://localhost:8888/api/v1/bundle_versions/?bundle__uuid=33566cb9-e607-425b-8cd1-db0ad4ab97d7
   * list BundleVersions for a Collection, e.g.
      http://localhost:8888/api/v1/bundle_versions?bundle_collection__uuid=4ffa83df-9b93-499e-91c29f5a7ad9
- [x] Add PUT method to `BundleVersionViewSet` to allow changing `change_description`: 023d251
   * Visit http://localhost:8888/api/v1/bundle_versions/ and view a bundle version.
   * Update the `change_description` field and submit.
- [x] Allow posting multiple files at once: 3c93aee
   * This change cannot be tested from the DRF GUI
   * See the [`test_create_multiple_files`](https://github.com/open-craft/blockstore/commit/3c93aee20a9dead5c5036b816861f1da0f28c203#diff-3959f0e2e005ddd8ad570fb6febdea94R431) unit test for details on how this feature works.
- [x] Verify the exact output in view tests: 6fdae38 

**Reviewers**
- [ ] @symbolist

bradenmacdonald commented:
Yay!

bradenmacdonald commented on requirements/base.txt:
```diff
@@ -1,5 +1,5 @@
 attrs<19.0.0
-django==1.11.14
+django==1.11.15
```
can we use `django>=1.11,<1.12` instead? I'm not clear why we would want to pin this to a minor version.

pomegranited replied:
My only reason to pin it was to [match edx-platform](https://github.com/edx/edx-platform/blob/master/requirements/edx/django.txt#L1), but yep, can change this.

bradenmacdonald replied:
Blockstore uses python 3 and has its own virtual environment so I don't think we need to match.

bradenmacdonald commented:
@pomegranited @symbolist @ormsbee 

This can be for a future PR, but I'd like your thoughts:

Currently, bundle files get serialized like this:

```
    {
        "data": "http://localhost:18250/media/09b3b612-3552-474a-bab9-4ebf7aebee22/data/2d97e22ef5e40f7b65eaf26d3169006739e75a7c",
        "path": "/static/handout.pdf",
        "public": true,
        "size": 15001,
        "url": "http://localhost:18250/api/v1/bundles/09b3b612-3552-474a-bab9-4ebf7aebee22/files//static/handout.pdf/"
    },
```
do we need a separate `data` and `url`? Wouldn't it be nicer to just have:
```
    {
        "path": "/static/handout.pdf",
        "public": true,
        "size": 15001,
        "url": "http://localhost:18250/api/v1/bundles/09b3b612-3552-474a-bab9-4ebf7aebee22/files/static/handout.pdf"
    },
```

Or if an endpoint to get info about a single file is important maybe we can add a `"metadata_url": "http://localhost:18250/api/v1/bundles/09b3b612-3552-474a-bab9-4ebf7aebee22/files/static/handout.pdf?metadata"` field?

Also, it would be much nicer if the data URLs were served with the correct content-type and filename. Currently whenever I try to view any blockstore file contents, it forces a download of a file with a long alphanumeric name and no file extension. Simple text based formats, PDFs, images, etc. should just open in the browser.

bradenmacdonald commented on blockstore/apps/api/v1/tests/test_views.py:
```diff
@@ -139,28 +177,45 @@ def test_list(self):
+            'version_num': self.bundle_version.version_num,
         })
-        response = self.client.get(url)
+        self.assertEqual(response.status_code, 200)
```
If you modify `self.response()` to have an optional `expected_response_code=200` parameter and put the `self.assertEqual(response.status_code, expected_response_code)` line inside that method, you can remove the repeated `self.assertEqual(response.status_code, 200)` from all these tests.

pomegranited replied:
Nice, thanks -- addressed with 206984e .

ormsbee commented:
In terms of the file name vs. its location, I had a couple of goals:

* File renames shouldn't be expensive in terms of space, particularly since we're never going to delete old versions.
* We should be able to serve content to the client without having to stream it through Django (for performance reasons).

For making links to hash-named data files that download with better file names, I use the `download` attribute on `<a href...>` tags (see Django Admin for Bundles). Later on, we can generate S3 links that have a [`Content-Disposition`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition) header set with the filename.

Given that, I think the friendliest thing we can do that keeps it operationally sane is:

1. Provide the `path` so a client knows what it's supposed to be named.
2. Provide a link to the raw data file. With S3, this link would be one that encodes the `Content-Disposition` header value to force the readable filename. We could make it so that links to the raw data for other storages types have a querystring param that we could interpret in middleware/nginx config to tack on the `Content-Disposition` header as well.
3. Make sure that `<a>` tags have the `download` attribute set.
4. Make a cleaner set of URLs that look like the filesystem would, and have it be a redirect to (2).

bradenmacdonald commented:
@ormsbee 

Ah, that all makes sense. I don't like how the aesthetics of how the "file" API URLs currently require a double slash and a trailing slash (`bundles/blah/files//file.txt/`), but otherwise I think that all makes sense.

> For making links to hash-named data files that download with better file names, I use the download attribute on <a href...> tags (see Django Admin for Bundles). 

Oh, that's nice! I actually hadn't even seen the django admin for bundles, so I didn't know it was so nicely set up. That's what I wanted to see :)

When working on the runtime that loads content out of blockstore (https://github.com/edx/edx-platform/pull/19136), I realized that we might want two data URLs for each file in the API, one being the private URL that can load the contents of any file in the bundle (for consumption by Studio and the LMS etc., for small data files) and the other being the public URL (sent directly to browsers, for asset files). (We can also make those the same and rely on blockstore clients to check the "public" attribute before publishing URLs, but I think that is more prone to bugs where blockstore clients implement it incorrectly.)

ormsbee commented:
@bradenmacdonald: Yeah, I'd prefer we lose the extra slashes as well.

> When working on the runtime that loads content out of blockstore (edx/edx-platform#19136), I realized that we might want two data URLs for each file in the API, one being the private URL that can load the contents of any file in the bundle (for consumption by Studio and the LMS etc., for small data files) and the other being the public URL (sent directly to browsers, for asset files). (We can also make those the same and rely on blockstore clients to check the "public" attribute before publishing URLs, but I think that is more prone to bugs where blockstore clients implement it incorrectly.)

Interesting. I like the idea of stripping public/private knowledge out of Blockstore in general, as it makes things simpler and allows clients to make make later-binding decisions on that. But I believe doing that puts us in a place where we have to either:

* default everything to private in the buckets themselves and make public links via relatively short lived signed URLs.
* copy the public files to some other store at publish time.

Does that sound right?

bradenmacdonald commented:
@ormsbee 

Yeah, exactly. At first I misunderstood what you just said, but now that I thought it through, it makes sense. I think the term "private/public" is a bit ambiguous:

We need to prevent learners from accessing raw `.olx` files, because they contain answers etc., but learners need direct access to asset files like images and PDFs, so we'll need something like:

* Within a bundle, all files are "private" by default, except those in some well-known prefix like `/public`
* All files have unguessable `data` URLs
* When retrieving any file's metadata, for most files, blockstore will send e.g. 
   ```json
    {
        "private_data_url": "https://s3-bucket/bundle/data/8f273087yq4n8c7qt89whu9o",
        "public_data_url": null
    }
    ```
* When retrieving any `/public` file's metadata, blockstore will send e.g. 
   ```json
    {
        "private_data_url": "https://s3-bucket/bundle/data/8f273087yq4n8c7qt89whu9o",
        "public_data_url": "https://cloudfront-distribution/bundle/data/8f273087yq4n8c7qt89whu9o"
    }
    ```
* On local devstacks, the private data URLs may differ from the public ones and can have the same API authentication as the rest of the blockstore API, to ensure they're not being served directly to users.

However, I realize there are two different uses of the word "private"; the other use is that "public" assets like PDF files still need to be restricted in some cases, e.g. so that only users who are enrolled in the course can download them. In that case, the decision shouldn't be made by blockstore, so it does sound like "default everything to private in the buckets themselves and make public links via relatively short lived signed URLs" yep. (Blockstore will always hand out signed public links for asset files, and the LMS can decide if the user is allowed to use that link or not).

ormsbee commented:
> @ormsbee 4. Hmm. Will Blockstore be serving these redirects? Seems quite expensive for assets.

I don't think the redirects will be that expensive, but we can definitely punt on implementing it for a while.

> Though I do think this info should be in `bundle.json` and `FileInfo` objects should only have storage relevant metadata (i.e. should not have the public field it currently has). We will likely have to denormalize the access metadata though (to avoid having to read bundle.json everytime) but there should be a separate plugin which reads new snapshots and does that?

Since it's immutable, we can cache it pretty aggressively on the client side. How large do you think that file could grow?

bradenmacdonald commented:
What's the advantage of maintaining an `assets` list in `bundle.json` compared to a convention of using a prefix like `/public/` for asset files?

pomegranited commented on Makefile:
```diff
@@ -18,7 +18,7 @@ help:
 	@echo ""
 
-VENV_PATH?=/blockstore/venv
+VENV_PATH?=$(or ${VIRTUAL_ENV},${VIRTUAL_ENV},/blockstore/venv)
```
Yep, I got that..  I'm not sure what kind of environment others are running Blockstore in, but I'm just using a simple `virtualenv`, and that sets the `VIRTUAL_ENV` variable, not a `VENV_PATH` variable.  And my virtualenv isn't in the default location.

symbolist replied:
Ah, is setting `VIRTUAL_ENV` the default behavior for `virtualenv`? If yes, we can always change `VENV_PATH` to it.

pomegranited replied:
Ok cool! I wasn't sure at all if that `VENV_PATH` was a critical name for someone's dev environment, and didn't want to break it for people.

But have replaced it with `VIRTUAL_ENV` for a cleaner diff here, cf f41d24f , https://github.com/open-craft/blockstore/pull/8/files#diff-b67911656ef5d18c4ae36cb6741b7965

pomegranited commented on blockstore/apps/api/v1/serializers/bundles.py:
```diff
@@ -26,6 +28,11 @@ class Meta:
             'versions',
         )
+        expandable_fields = {
+            'files': (ExpandedFileInfoField, (), dict(
```
Oo nope, there's no pagination of any kind on blockstore yet, and none in this expansion either.

symbolist replied:
Okay, I don't think there should be any here either.

pomegranited commented on blockstore/apps/api/v1/views/bundles.py:
```diff
@@ -33,6 +37,9 @@ class BundleVersionViewSet(viewsets.ReadOnlyModelViewSet):
     lookup_value_regexes = (UUID4_REGEX, VERSION_NUM_REGEX)
 
+    filter_backends = (DjangoFilterBackend,)
+    filterset_fields = ('bundle__uuid',)
```
Sure thing -- done with 2565f77 .

pomegranited commented on blockstore/apps/api/v1/tests/test_views.py:
```diff
@@ -220,171 +317,238 @@ class BundleFileReadOnlyViewSetTestCase(ViewsBaseTestCase):
+            },
+            method='post',
+            data={
+                'data': [ContentFile(TEXT_CONTENT_BYTES), ContentFile(HTML_CONTENT_BYTES)],
```
> 1. Isn't this going to require base64 encoding binary files?

Nope, no change to the encoding requirements, cf https://github.com/open-craft/openedx-blockstore-relay/pull/5

> 2. This is not a very standard REST-ful payload format.

I found a few discussions about base64/JSON vs multipart ([1](https://stackoverflow.com/questions/33279153/rest-api-file-ie-images-processing-best-practices), [2](https://github.com/json-api/json-api/issues/246), [3](https://stackoverflow.com/questions/26984844/multipart-or-base64-for-ajax-file-uploads)).  Multipart uploads can be streamed, and aren't as large as base64-encoded requests, and don't need to be decoded all at once in order to be processed.

Would you rather it was base64 encoded and JSON-ified instead?  I could make that change if needed.

symbolist replied:
Ah okay! I went and looked at the [encoders code](https://github.com/requests/requests/blob/master/requests/models.py#L110-L169) and that makes sense now. I cannot think of any technical issue with this option. [DRF](https://github.com/edx/django-rest-framework/blob/master/rest_framework/parsers.py#L110) does seem to read the whole stream to create the `data` dict so to benefit from that, we will have to do some more work but its an optimization we can always do later on when needed. With JSON that will not be possible at all.

The only other non-ideal thing is the shape of the data payload. But does that matter? 

Having said so this should work for the needs of the drafts API so lets see what @ormsbee has to say.

P.S. https://tools.ietf.org/html/rfc7578 is a fun read.

ormsbee replied:
Works for me. :)

pomegranited commented on blockstore/apps/api/utils.py:
```diff
@@ -0,0 +1,20 @@
+    def zip(self):
+        """
+        Returns a list of dicts using the keys and values lists.
+        """
```
Sure thing -- is 33cb20d sufficient?

pomegranited commented on blockstore/apps/api/v1/serializers/snapshots.py:
```diff
@@ -49,11 +51,81 @@ class FileInfoSerializer(serializers.Serializer):
+        if isinstance(instance, BundleVersion):
+            bundle_version = instance
+        else:
+            bundle_version = instance.get_bundle_version()
```
Done with 2565f77.

pomegranited commented on blockstore/apps/api/v1/serializers/snapshots.py:
```diff
@@ -49,11 +51,81 @@ class FileInfoSerializer(serializers.Serializer):
+        self.view_name = view_name
+
+        # Ignore the context argument if passed from the ExpanderSerializerMixin
+        kwargs.pop('context', None)
```
Explained with 2565f77.

pomegranited commented:
Thanks for your reviews @symbolist and @ormsbee !

I've addressed all of your comments, and left "unresolved" the ones where I'm uncertain about next steps.  Let me know if there's anything else I should do here?

pomegranited commented:
Thanks @symbolist , I've addressed your remaining comments.  Could you approve if you're happy?

@ormsbee  could you please let us know if you're ok with the [mime/multipart multiple file uploads payload implemented here|https://github.com/open-craft/blockstore/pull/8#discussion_r228184313], and I think we're good to merge?

symbolist commented:
@ormsbee 

>> 4. Hmm. Will Blockstore be serving these redirects? Seems quite expensive for assets.
> I don't think the redirects will be that expensive, but we can definitely punt on implementing it for a while.

To be clear, I meant for the clients they add latency. But that depends on whether end-users will be accessing those URLs from browsers/apps or are they just for consumption by LMS/Studio.

>> Though I do think this info should be in bundle.json and FileInfo objects should only have storage relevant metadata (i.e. should not have the public field it currently has). We will likely have to denormalize the access metadata though (to avoid having to read bundle.json everytime) but there should be a separate plugin which reads new snapshots and does that?

> Since it's immutable, we can cache it pretty aggressively on the client side. How large do you think that file could grow?

Ah, I see. Can't see why it should be more than a couple of KB for most bundles?

pomegranited commented:
Ah oops, thanks for catching that @symbolist , fixed with f549b3a.

I've got two 👍 s here, so merging now :)  Thanks for your reviews @symbolist  and @ormsbee !

edx-webhook commented:
Although this pull request is already merged, I've created [OSPR-5481](https://openedx.atlassian.net/browse/OSPR-5481) so that we can track it in Jira.

There is nothing you have to do. No action is needed from your side.  Thanks again for your contribution.

<!-- comment:welcome_closed -->
<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->
